File: src/main/java/com/mutableconst/sql/util/SqlUtils.java
package com.mutableconst.sql.util;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;

public class SqlUtils {
    private static final Pattern namedParamRegex = Pattern.compile("(?<!:):([A-Za-z]+)");

    public static String unparameterizeSql(String preparedSql) {
        Matcher namedSqlMatcher = namedParamRegex.matcher(preparedSql);
        String unparameterizedSql = namedSqlMatcher.replaceAll("?");
        return unparameterizedSql;
    }

    public static Map<Integer, String> extractParameters(String preparedSql) {
        Map<Integer, String> parameterMap = Maps.newHashMap();
        Matcher namedParamMatcher = namedParamRegex.matcher(preparedSql);
        int currentParamIndex = 1;
        while(namedParamMatcher.find()) {
            parameterMap.put(currentParamIndex, namedParamMatcher.group(1));
            currentParamIndex++;
        }
        return ImmutableMap.<Integer, String>builder()
                .putAll(parameterMap)
                .build();
    }
}

File: src/main/kotlin/com/mutableconst/sql/util/SqlUtils.kt
package com.mutableconst.sql.util

import com.google.common.collect.ImmutableMap
import com.google.common.collect.Maps
import java.util.regex.Pattern

object SqlUtils {
    private val namedParamRegex = Pattern.compile("(?<!:):([A-Za-z]+)")

    fun unparameterizeSql(preparedSql: String): String {
        val namedSqlMatcher = namedParamRegex.matcher(preparedSql)
        val unparameterizedSql = namedSqlMatcher.replaceAll("?")
        return unparameterizedSql
    }

    fun extractParameters(preparedSql: String): Map<Int, String> {
        val parameterMap = Maps.newHashMap<Int, String>()
        val namedParamMatcher = namedParamRegex.matcher(preparedSql)
        var currentParamIndex = 1
        while (namedParamMatcher.find()) {
            parameterMap.put(currentParamIndex, namedParamMatcher.group(1))
            currentParamIndex++
        }
        return ImmutableMap.builder<Int, String>().putAll(parameterMap).build()
    }
}
